Truncate data written to logs in C.I. environment

jugglinmike opened the pull request:
My choice of maximum row count is completely arbitrary. I'm happy to change it to any other value based on the experience of my reviewers. For context:

- the log generated for the patch that identified this issue (gh-4549) triggered tests for 15 "noisy" files and produced a log that exceeded Travis CI's 4 megabyte limit.
- the log generated for this pull request (which modifies the same 15 files) is 1.1 megabytes in size.

Please note that this solution (limiting overall log size by truncating per-file tables) does *not* scale for patches that effect a large number of tests that contain many subtests. In those cases, the overall log size may grow beyond TravisCI's 4MB limit despite the truncation of each individual test's table. Supporting this case gracefully would significantly increase the script's complexity (e.g.  log buffering techniques, row allocation heuristics, etc.). This is likely not warranted given the fact that such patches are equally onerous for reviewers to validate and are likely discouraged as a matter of policy.

Further truncate the log posted to GitHub.com. This is a hard clamp at GitHub's maximum length, minus the text necessary to notify the reader that the truncation occurred. This operation may invalidate part of the output (e.g. Markdown tables of open HTML tags), but again, I think we should prefer simplicity to completeness in guarding against this particular edge case. A link to the Travis "build job" page (containing a "less-abbreviated" version of the output) is included where available.

If all that seems reasonable, then this should resolve gh-4550.

jugglinmike committed:
Truncate data written to logs in C.I. environment

jgraham commented on check_stability.py:
```diff
@@ -33,6 +33,7 @@
 wptrunner_root = None
 
 logger = logging.getLogger(os.path.splitext(__file__)[0])
+MAX_TABLE_ROWS = 256 if "CONTINUOUS_INTEGRATION" in os.environ else float("inf")
```
Should be `else None` I think.

jugglinmike replied:
Python appears to interpret `None` as being "less than" all other number values:

    >>> [(None < x) for x in [float('inf'), 1, 0, -1, -float('inf')]]
    [True, True, True, True, True]

Since we don't need to enforce a log size limit outside of continuous integration, I don't think this appropriate for use as a maximum value.

jgraham commented on check_stability.py:
```diff
@@ -170,16 +173,31 @@ def get(self, url, headers=None):
+                "excess of GitHub.com's limit for comments (%s characters)." +
+                "%s*\n\n") % (len(body), self.max_comment_length, link)
+
+            body = truncation_msg + \
```
Avoid line continuation characters.

jugglinmike replied:
Got it. (I'm making it obvious that I rarely write Python, aren't I?)

jgraham commented on check_stability.py:
```diff
@@ -211,7 +229,15 @@ def emit(self, record):
+        travis_job_id = os.environ.get("TRAVIS_JOB_ID")
+
+        if travis_job_id is not None:
+            full_log_url = \
```
Again, please remove the line continuation character.

jgraham reviewed (commented):
It seems like a more sophisticated approach would be to count the number of bytes in `log` and allow 3Mb of output from the python script in travis before printing a truncation message and then making `log()` a noop.

tobie commented:
Ping @wpt-pr-bot.

jugglinmike commented:
@jgraham I've pushed up a commit that implements the solution at a lower level (the "Python file descriptor" level, but not so low as the "process output stream" level). I've defined it in terms of a new command-line option for the stability script--maybe a little over-engineered, but I wanted to make the behavior as discoverable as possible.

Anyway, [the latest build logs](https://api.travis-ci.org/jobs/198201110/log.txt?deansi=true) demonstrate the intended behavior:

>     INFO:check_stability:| `Range 7 [paras[1].firstChild, 2, paras[1].firstChild, 9] followed by Range 2 [paras[0].firstChild, 2, paras[0].firstChild, 8]: first addRange() must result in the selection's last range being the same object we added`                 | PASS    |
>     INFO:check_stability:| `Range 4 [paras[1].firstChild, 0, paras[1].firstChild, 0] followed by Range 21 [foreignDoc.body, 0, foreignDoc.body, 0]: first addRange() must result in the selection's last range being the same object we added`                        | PASS    |
>     INFO:check_stability:| `Range 6 [paras[1].firstChild, 2, paras[1].firs
>     Log reached capacity (3145728 bytes); output disabled.
>
>
>     The command "bash $SCRIPT" exited with 0.
>     store build cache
>
>     change detected (content changed, file is created, or file is deleted):
>     /home/travis/.cache/pip/http/0/1/e/c/5/01ec55882d7fee6cccf9bafb035721d3b0f1c4f5d70ef515794d5153
>     /home/travis/.cache/pip/http/0/4/6/f/2/046f263a04464ad173f2b55b53089d6611b281fd92901c6030b4cd93
>     /home/travis/.cache/pip/http/0/8/d/c/e/08dcef3be0bb9cce46e42c11d622ed3fea0797e37de9ffac3f5d70d7
>     /home/travis/.cache/pip/http/0/e/6/0/8/0e6088bd8e09b5becb00f7b07c4c7e9f5ce58d23a05bb02f89d475b4
>     /home/travis/.cache/pip/http/2/7/6/b/0/276b09a4fa1b1ca2bb4fd4b43f5c3d25e8d0fea3c8af6edde614d8df
>     /home/travis/.cache/pip/http/2/7/e/3/e/27e3e8f32f819b0843f464a41afb5e987e18f8a89ab01ac2ad98f8b0
>     /home/travis/.cache/pip/http/2/d/9/f/0/2d9f0780aedbe51e7cdcb37bf4c03c32f9d05e9445fc2fb4bf91af6c
>     /home/travis/.cache/pip/http/3/8/1/8/e/3818e6cf419a275375f4fd69142c6bcb13c625d3d0a0bd5eecb07057
>     /home/travis/.cache/pip/http/3/a/f/3/a/3af3addf06e983a6c02f46e7bea70c221d3ff95bf1418fa6da354e14
>     /home/travis/.cache/pip/http/4/7/9/9/6/479960059f6a2c1bdbd0c32805cce11c8c453c4dcac2ea743c8462ff
>     /home/travis/.cache/pip/http/5/c/d/e/b/5c
>
>     ...
>
>     changes detected, packing new archive
>     uploading archive
>
>
>     Done. Your build exited with 0.

What do you think?

jugglinmike commented:
By the way: the modifications to the "selection" tests are a red herring. I've included them to demonstrate the change to the stability bot's behavior, but they won't actually be merged when this patch is landed.

jugglinmike commented:
Part of @bobholt's work in gh-4651 involved the removal of GitHub-specific logic from the stability bot. Because this patch modified the removed logic, this introduced merge conflicts here. I've resolved the conflicts and force-pushed the new version of the change set. (The original version of this branch is available on my fork for posterity: https://github.com/w3c/web-platform-tests/compare/master...jugglinmike:4550-limit-table-size-prebase.) I have ported the GitHub-specific functionality to [the new "prbuildbot" project](https://github.com/w3c/prbuildbot); [that change is under review here](https://github.com/w3c/prbuildbot/pull/2).

jugglinmike commented on check_stability.py:
```diff
@@ -107,6 +106,48 @@ def __exit__(self, type, value, traceback):
         print("travis_fold:end:%s" % self.name, file=sys.stderr)
 
 
+class FilteredIO(StringIO):
```
I've been researching how to do this more completely, and I found the following
documentation:

> `stdout` and `stderr` needn’t be built-in file objects: any object is
> acceptable as long as it has a `write()` method that takes a string argument.

(https://docs.python.org/2/library/sys.html#sys.stdout)

It seems like for this use case, the naive "write-only" implementation is
sufficient. Would you agree?

jugglinmike replied:
I think I follow you. I've pushed up a commit that makes the class inherit from "object" and implements the `__getattr__` method as a proxy. Let me know if this is what you had in mind.

jugglinmike commented on check_stability.py:
```diff
@@ -582,6 +622,11 @@ def get_parser():
                         # This is a workaround to get what should be the same value
                         default=os.environ.get("TRAVIS_REPO_SLUG").split('/')[0],
                         help="Travis user name")
+    parser.add_argument("--output-cap",
```
Sure (though I'll point out that this increases the noun/verb ambiguity in the word "output")

jugglinmike replied:
Addressed this in 1501eac729102ad130a40d0f9d3120bdd49312ff. GitHub seems a little slow today, so you may have missed that

jugglinmike commented on check_stability.py:
```diff
@@ -582,6 +622,11 @@ def get_parser():
                         help="Travis user name")
+    parser.add_argument("--output-cap",
+                        action="store",
+                        default=sys.maxint,
```
Using `None` would force us to opt-out of argparse's type enforcement. What do you think about using `-1` to signify "no limit?"

jgraham commented on check_stability.py:
```diff
@@ -592,11 +637,18 @@ def main():
     parser = get_parser()
     args = parser.parse_args()
 
+    replace_streams(args.output_bytes,
```
`if args.output_bytes is not None:` perhaps?

jugglinmike replied:
Looks like I can support this by simply not specifying a "default" value when defining the option.

jgraham reviewed (approved):
Please remove the test changes.

jgraham commented on ci_stability.sh:
```diff
@@ -44,7 +44,7 @@ install_chrome() {
 
 test_stability() {
-    python check_stability.py $PRODUCT
+    python check_stability.py $PRODUCT --output-cap $((1024 * 1024 * 3))
```
This should be `--output-bytes`

w3c-bots commented:
# Chrome (unstable channel) #
Testing web-platform-tests at revision d716371c7823185ddaf668138a3a7caf76eb2a5f
Using browser at version 58.0.3004.3 dev
Starting 10 test iterations
No tests run.

w3c-bots commented:
# Firefox (nightly channel) #
Testing web-platform-tests at revision d716371c7823185ddaf668138a3a7caf76eb2a5f
Using browser at version BuildID 20170123125947; SourceStamp 36486fdc3813ef7943ae5b07b4128866d1938a6c
Starting 10 test iterations
No tests run.

jugglinmike commented:
Updated the command-line flag usage, rebased onto master, squashed, and removed the "demo" commit. The prior version of this branch is available here: https://github.com/w3c/web-platform-tests/compare/master...jugglinmike:4550-limit-table-size-prebase-2

Thanks for the review, James!
